### ISSUE TITLE
Instance Opt-in

### ISSUE THREAD

**Issue by h-2** [snarfed/bridgy-fed#1305]
I know that you got a lot of heat initially when starting the bridge as opt-out. I would still like you to reconsider this, or at least allow entire Mastodon instances to opt-in for their users. This would allow us to lobby the big instance admins on both sides to opt-in, while ignoring nerd and niche instances.

I really like cross-federation. In fact, I think it's probably one of the most important features we need right now. However, at the moment this feature is absolutely niche. On mastodon.social the bridge account has less than a 1000 followers ( 0.4 % of active users), and on bsky it has around 6000 followers (0.075% of total users). BUT for proper interaction users on both sides need to have opted in, so chances of two random users being connected are tiny 😢 

Have you considered instance opt-in?

P.S.: is there a way to support this project with money? Patreon, Paypal, Github, SEPA?

**Comment by ygg2** [comment]
I think it wouldn't be good to create a situation where admins can do it aginst their users' wishes, potentially users who don't want to be bridged but haven't put #nobridge in their profiles because it's opt-in. There's already a lot of frustration over admins being the determiners of who an instance federates with (both in blocking and not blocking) especially without much transparency and the difficulties of migrating away as a user if you don't like the decision. It might also be kinda one-sided since Bluesky is more of a single user experience without really having PDS admins in the way fedi instances do

https://github.com/snarfed/bridgy-fed/issues/966 should help a lot with opt-in though imo

**Comment by qazmlp** [comment]
There's some previous discussion about this: https://github.com/snarfed/bridgy-fed/discussions/1118

**Comment by h-2** [comment]
> I think it wouldn't be good to create a situation where admins can do it aginst their users' wishes, potentially users who don't want to be bridged but haven't put #nobridge in their profiles because it's opt-in.

TBH, fediverse culture seems to be that you should just go to whatever instance where you like the rules, so I don't think this makes a difference one way or another.
Users can always block an instance individually. Or they can join another instance if they don't like the defaults on their instance. Some people did this when mastodon.social started federating with threads, but 99% of users couldn't care less.

Anybody can start their own Mastodon server and post the most terrible content. Federation with that server would still be opt-out (on instance basis or individual basis).
So there is no logical explanation for people being upset about bridgy being opt-out. The only difference is a different protocol/tech.

So I still think this should be pursued!

**Comment by qazmlp** [comment]
There's definitely a difference in baseline consent between ActivityPub and ATProto.

Both user-discovery and search-indexing of posts are mostly opt-in on AP, while they're effectively mandatory on ATProto.
*From a purely technical point of view* this isn't enforced in ActivityPub of course (since that's not technologically feasible), but ignoring these preferences is still a social boundary violation that regularly leads to services being widely defederated.

I think the most recent example in Newsflip, who implemented following of ActivityPub actors for their accounts without federating those profiles. They tried to get around not receiving content by default by mass-following from a service account, which got `newsflip.com` completely blocked on many large instances. (The instance where my main account is suspended `newsflip.com` and limited `newsflip.social` to the point I can't see posts from there at all unless I follow the author.)

---

That said, it's very much true that baseline consent on individual instances can differ from the protocol at large, so the feature does make sense to me. Instance admins usually do have the tooling to poll their users and establish consent, as well as to notify about policy changes ahead of time.

To avoid misunderstandings at the expense of Bridgy Fed, the bridged bio on other networks should clearly say that the account is bridged by default because the instance opted in though.

**Comment by snarfed** [comment]
Thanks for the discussion, all! And thanks @qazmlp for linking to https://github.com/snarfed/bridgy-fed/discussions/1118 . We don't really need to reopen the whole opt in vs opt out debate in general, but @h-2 you're definitely not the only person who wants this, a number of other fediverse instance admins have asked for it too.

Sounds like there's at least some rough agreement here, at least between you all: it's probably ok for instances to automatically opt their users into the bridge, _if_ they warn people loudly beforehand, during signup, _and if_ users can easily opt out.

As @h-2 mentioned, this feels similar to federation in general in the fediverse. Federation with other instances is usually open by default, and users and admins can both block/defederate specific instances if they want.

---

Concretely, here's how this would work, from https://github.com/snarfed/bridgy-fed/discussions/1118#discussioncomment-9693007 :

> One catch is that right now, at least the way fediverse software is generally designed, I'd still need to have the @bsky.brid.gy@bsky.brid.gy bot user follow each user on an auto-enabled instance like this, so that it delivers their posts to BF. So, they need some way to automatically have every new user follow the bot user, so that it will follow them back. It sounds like this is maybe possible in at least some software, eg Mastodon...
>
> ...but if that's necessary, and they do that, then there's nothing special I'd need to do on BF's end.

**Comment by snarfed** [comment]
@h-2 if you or any other Mastodon admin set this auto-following up, please let me know how it goes, I'd love to hear!

**Comment by qazmlp** [comment]
I'm pretty sure that wouldn't work as-is. Mastodon accounts start out without profile picture or bio, and they'd also not meet the account age requirement.

**Comment by snarfed** [comment]
Good points!

**Comment by snarfed** [comment]
When an instance is ready to try this, I can exempt them from the spam filter req'ts.

**Comment by h-2** [comment]
> @h-2 if you or any other Mastodon admin set this auto-following up, please let me know how it goes, I'd love to hear!

I don't run my own instance, at least not at the moment. I am more interested in this from a user-perspective, i.e. I would try to convince the admins of the instances that I am on to activate such a feature.
The most important aspect from my POV is creating critical mass. Unless I can assume that *most* users on one network are able to interact with *most* users on the other network, it remains a niche feature. And tbh, repeatedly checking my mirrored profile page on BSky, checking if someone followed me there, then checking whether that person is not following me on Mastodon, then sending a message to that handle on the bridge to convince them to follow the bridge's handle on BSky... well it's just too much work to be viable in practice.

It's also, why I think the "follow the bridge account by default" is not bad, but it's also not what's going to bring about major change, since I don't imagine any instance retro-actively changing this for all their existing users.

For transparency: I recently opened this issue on the main Mastodon issue tracker, but have received only one (not very helpful) answer to date:
https://github.com/mastodon/mastodon/issues/31771

**Comment by qazmlp** [comment]
For what it's worth, there's still that "relay" method I mentioned to get specifically public posts in a lightweight way instance-wide, but that would definitely require quite a bit of setup on the Bridgy Fed side.

**Comment by snarfed** [comment]
@qazmlp yes! Do you know of any good technical docs on relays? What the protocol and format are, which activities are expected - ie is it just `Create`s for posts, or also `Update`/`Delete`, user profile `Update`s, etc? I've found lots of high level docs, and how to add them as a Mastodon admin, and projects, but nothing technical enough to build against.

**Comment by qazmlp** [comment]
> @qazmlp yes! Do you know of any good technical docs on relays? What the protocol and format are, which activities are expected - ie is it just `Create`s for posts, or also `Update`/`Delete`, user profile `Update`s, etc? I've found lots of high level docs, and how to add them as a Mastodon admin, and projects, but nothing technical enough to build against.

I still can't find any official technical documentation unfortunately, but the gist is that relays are inboxes where you can `Follow` the `…#Public` collection: https://github.com/mastodon/mastodon/blob/592a7af27f7699d4751d2bea7785149d3c0e5d58/app/models/relay.rb#L48-L56
Only an authorised user (admin) can create this kind of Follow activity, so this should be enough for authentication.

The relay doesn't need to follow back, I think, though Aode relay will do so if the instance actor is followed directly: https://git.asonix.dog/asonix/relay/src/commit/a23b30cc91cad42a67c9c8d31ae76f946cca4969/src/jobs/apub/follow.rs

After the initial follow is accepted, the instance will send public activity to that inbox. This could differ by software, but from Mastodon you will get at least `Update`s and `Delete`s for Public posts (along with `Create`s), as well as any `Update`s and `Delete`s for accounts:
https://github.com/mastodon/mastodon/blob/592a7af27f7699d4751d2bea7785149d3c0e5d58/app/models/relay.rb#L48-L56.
https://github.com/mastodon/mastodon/blob/592a7af27f7699d4751d2bea7785149d3c0e5d58/app/models/relay.rb#L48-L56

I'd suggest to just use your general inbox and handle everything as normal (i.e. with the usual filters), aside from a special case for that `…#Public` `Follow` and `Undo`-`Follow`.

At this point the relay could also start sending public activity to the follower, but for Bridgy Fed I definitely wouldn't do that outside hashtag- or Bluesky-feed-based opt-ins for example.

**Comment by qazmlp** [comment]
Yeah, it's unfortunate that there's such a mismatch in expectations.

*Arguably*, if an ActivityPub user has all of

```json
{
  "https://www.w3.org/ns/activitystreams#manuallyApprovesFollowers": false,
  "http://joinmastodon.org/ns#discoverable": true,
  "http://joinmastodon.org/ns#indexable": true
}
```

_and_ Bluesky as a whole was opt-out (so that all followers are visible and can be blocked if necessary), then that _could_ reasonably be interpreted as Bluesky/ATProto-compatible as long as the bridged profile requests to not be web-visible, but I'm sure there'd still be people who'd complain about it quite fiercely even then.

(At least `indexable` is opt-in, not sure if `discoverable` is. `indexable` may be younger than Bridgy Fed, not sure.)

**Comment by Tamschi** [comment]
I'll reply to https://github.com/snarfed/bridgy-fed/issues/1471#issuecomment-2474092931 here since I think this post would be off-topic in #1471.

---

> […] (and we are at 275M monthly users anyways).

We should _definitely_ look into letting instances provide ATProto handles for their users by default alongside this feature here.

@pcottle In case you're not aware, one of the hard caps Bridgy Fed has with its current infrastructure is how many ATProto handles it can provide at once (#744). As of a few days ago, that was at 28k/50k used, and while increasing it doesn't seem to be too much of a problem right now, I think it's likely you'd exceed that very quickly if you enabled the bridge by default for everyone who has federation enabled.

I'm sure you'd also prefer your users to have `@….threads.net` handles rather than appearing as `@….threads.net.ap.brid.gy` handles on Bluesky.

@snarfed I think we can assume that the account that performs the instance opt-in is also authorised to configure handle defaults for the domain via DM in some way, but Bridgy Fed would also need a way to inform an instance of the relevant `@id`/DID/handle mapping changes whenever one occurs so that they can update their DNS and/or `.well-known` setup accordingly.

I'll sketch two options that are relatively on-protocol, if that's okay:

We could DM them to the service account as plaintext separately, but I think a much better solution would be to attach that information as extra properties to the `Create`-`Note` (DM) or `Accept`-(/`Undo`-`Accept`-)`Follow` activities we send towards the users on those occasions. (Even if Threads doesn't support AP DMs properly, they could still easily handle those statuses internally.)

Alternatively, we could send separate activities to users just for DID-and-handle updates, which to me seems the cleanest overall but would require most original semantics.

The off-protocol option would probably be some kind of webhook call, but that seems like it would require additional interactive setup.

**Comment by pcottle** [comment]
I got a mini brief on the 28k/50k handle issue, but I figured we only take up those slots as either people opt-in (in the case of opt-in) or people get followed/followed by in the case of opt-out (which would be a much smaller number).

But you're right, if a lot of the tech threads folks start jumping onto the bluesky bridge experience, we might hit that limit.

What are the steps for serving AT handle resolution on our domain but not having to host an entire AT frontend? Especially if its just the steps associated with DID:
https://www.w3.org/TR/did-core/

it would be easy for us to defend that work

**Comment by Tamschi** [comment]
Bluesky users can't natively look up a remote account that's not present on their network yet, unfortunately.

In theory it would be possible to integrate it into the existing request flow though, i.e. when someone requests a user on an opt-out instance (via Chat message to `@ap.brid.gy`), it would announce the account to ATProto only then and reply that the account is now bridged due to the instance opt-in. I think that would be strange behaviour for smaller instances, but it could make sense for Threads.

Bridgy Fed uses `did:plc:…` rather than `did:web:…` in order to properly support migrations (#1137), and it takes care of the DID management for bridge accounts automatically.

On your end, you'd only have to set up DNS or serve the `did:plc:…` string as `text/plain` at `https://${handle}/.well-known/atproto-did` for each account so that AppViews can validate the handles. (See near the end of [How to set your domain as your handle - Bluesky](https://bsky.social/about/blog/4-28-2023-domain-handle-tutorial))

**Comment by snarfed** [comment]
Thank you @Tamschi! @pcottle, @Tamschi knows the current Bridgy Fed architecture well, all the info here so far is spot on.

> Bluesky users can't natively look up a remote account that's not present on their network already, unfortunately.

This is the key open question for instance-level opt-in right now. In the fediverse, user discovery is pull-based. When you search for someone's handle, your instance/client queries their webfinger and fetches their actor on demand. On Bluesky, though, we'd have to create a Threads user's bridged Bluesky account and profile ahead of time to make it visible.

We could do that! And even start bridging their posts immediately too. That's the most aggressive form of default-on bridging, which would generate the most load, and pressure on DNS capacity (#774), and also probably worry the most people from a privacy perspective.

The less aggressive approach, as @Tamschi mentioned, would be to bridge a Threads user only on demand, when a Bluesky user [requests them](https://fed.brid.gy/docs#dm-request). Safer and more conservative, but not as good for discoverability.

I'm open to either. The first (proactive) way would be a huge increase in load, though, at least assuming we start bridging osts proactively too, so I probably wouldn't want to do that [until we have an org and (probably) funding in place](https://snarfed.org/2024-10-30_53932).

**Comment by pcottle** [comment]
I know discoverability is worse for the on-demand model but I think that's the right downside to choose. In my mind the networks would start to merge like so:

- bridge is on, Threads is opt-out, Bluesky is either opt-in or opt-out
- the bridged Bluesky users start getting follows from Threads users organically once thats available
- A follow to any bridged account means we make the Threads account on-demand (so the *engaged* and tech-centric threads profiles start getting created, aka the people who would benefit from the bridge)
- Through normal account discovery (media recs, reposts, mentions and replies, etc), those Threads users get follows back from Bluesky (at least from bridged users, ideally from everyone on bluesky).
- Chill vibes as people can now discuss!

Does that make sense? If threads serves our own DIDs, does that resolve the DNS capacity concerns?

**Comment by Tamschi** [comment]
I think it makes sense for Bridgy Fed to first automatically push an opt-out-instance's ActivityPub user to ATProto when they `Follow` a Bluesky user, if needed. We'd want that to avoid #1102.

The DID documents themselves should be served by Bridgy Fed, in my eyes. That saves quite a lot of special-casing on our end.
If you serve just the handle verifications for your users, that alone avoids any issues with DNS capacity on our end.

What's the regex for Threads usernames? ATProto handles have to be DNS-compatible: https://github.com/snarfed/bridgy-fed/issues/982
If your usernames don't contain `.`, then I *think* you can use a single wildcard DNS entry and wildcard cert pair and only hit your database in the `GET /.well-known/atproto-did` handler.

I think both active notifications and a Bridgy Fed API to look up the DID for a bridge handle are feasible. The latter has the advantage that it avoids timing issues ("⚠️Invalid Handle") and doesn't need explicit recovery, but it would still be good to cache it for a bit. (It's really too little data to use an `ETag` or such, but we can send `Cache-Control: public, max-age=…, stale-while-revalidate=…` if that expires before a Threads handle can be reused by someone else. Even if not, the worst failure case would be seeing "⚠️Invalid Handle" temporarily on the new account reusing the name. We could of course also include the ActivityPub `@id` in the response for you to cross-reference.)

There are a few general edge cases we'd need to decide on, also involving interactions with other instances:

- If Bridgy Fed receives a bridged mention of a not-yet-bridged opt-out user, should that first bridge that user?

  This is probably fairly easy to do either way around since pushing users to AT is idempotent, but *not* bridging them is affected by #1163.

- What do we do for opted-in Public replies to opt-out users' posts?

  With the current architecture, it's definitely easier not to bridge either post, as-is. I think that bridging earlier posts lazily could also result in a confusing sparse and non-chronological timeline on their Bluesky profile, since the site rarely actually orders by post date.

  (Natively on ActivityPub, seeing a reply usually pulls in the conversation up to its root step-by-step, but it's difficult to do this in ATProto since each post in a conversation _has_ to mention the root's record key directly.)

**Comment by pcottle** [comment]
> What's the regex for Threads usernames?

its
```
(?P<username>[a-zA-Z0-9_]+(?:\\.[a-zA-Z0-9_]+)*)
```

so we definitely allow periods :)

I'm also fine serving just the handle verifications, whatever avoids the DNS issues!

**Comment by Tamschi** [comment]
I assume you have the infrastructure to run a custom DNS server 😅

I think it's easiest if you serve `_atproto.<handle>.threads.net` `TXT` records dynamically then, but the principle is the same as for `.well-known`. You'd ideally first check if the user exists on your end, then with Bridgy Fed to get the DID string.

For reference, here is the ATProto handle regex:

> `/^([a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?$/`

The underscores are replaced with `-` by Bridgy Fed already, so there's no issue with those since you don't allow dashes or tildes.
That said, leading or trailing underscores and those adjacent to a `.` are currently not supported.

**Comment by pcottle** [comment]
Serving the `.well-known` might be easier for internal execution at Meta (given the systems involved) -- its possible to take that approach still right?

**Comment by Tamschi** [comment]
Yes, I think so, but with some caveats.

In theory you may be able to use a wildcard entry for the entire DNS zone, but a cursory check tells me that could get a bit implementation-defined. It seems for wildcard TLS certificates to be fully compliant, you'll have to use a separate one for each level of nesting (But e.g. `*.*.threads.net` or `*.*.*.threads.net` should work everywhere, as long as each is the only entry in the respective certificate. Some browsers at least seem to be a bit more lenient, too).

**Comment by michael-patchwork** [comment]
> Thank you @Tamschi! @pcottle, @Tamschi knows the current Bridgy Fed architecture well, all the info here so far is spot on.
> 
> 
> 
> > Bluesky users can't natively look up a remote account that's not present on their network already, unfortunately.
> 
> 
> 
> This is the key open question for instance-level opt-in right now. In the fediverse, user discovery is pull-based. When you search for someone's handle, your instance/client queries their webfinger and fetches their actor on demand. On Bluesky, though, we'd have to create a Threads user's bridged Bluesky account and profile ahead of time to make it visible.
> 
> 
> 
> We could do that! And even start bridging their posts immediately too. That's the most aggressive form of default-on bridging, which would generate the most load, and pressure on DNS capacity (#774), and also probably worry the most people from a privacy perspective.
> 
> 
> 
> The less aggressive approach, as @Tamschi mentioned, would be to bridge a Threads user only on demand, when a Bluesky user [requests them](https://fed.brid.gy/docs#dm-request). Safer and more conservative, not as good for discoverability.
> 
> 
> 
> I'm open to both. The first (proactive) way would be a huge increase in load, though, at least assuming we start bridging posts proactively too, so I probably wouldn't do it [until we have an org and (probably) funding in place](https://snarfed.org/2024-10-30_53932).

I can see that for Threads on demand is the way to go but for other Fediverse instances it would be good to offer the proactive approach, combined with a default mechanism for the instance to give users their own ATProto handles as @Tamschi suggests - the two go together. We have something about to go live which I'll message about

**Comment by Tamschi** [comment]
Regarding https://github.com/snarfed/bridgy-fed/issues/1305#issuecomment-2476757067:

I looked into what's allowed in TLS certificates a bit more and `*.*….` is definitely _**not**_ allowed unless the rules changed recently. You could still do this if you have your own certificate authority to issue certificates dynamically (or equivalent), but otherwise it would be very difficult to go the `.well-known`-route while you have `.` in handles.

**Comment by pcottle** [comment]
I think converting `.`'s to some other character isnt the worst compromise. We just need to find an alternate (maybe `-`) that doesnt conflict with existing usernames :)

**Comment by snarfed** [comment]
Hah yes, we went through this whole arc ourselves, https://github.com/snarfed/bridgy-fed/issues/830

In the end, we couldn't quite get there because we needed to at least try to support all of the different fediverse software out there, but the problem space is obviously more manageable for you all with just a single, centrally managed platform. Looking forward to seeing where you end up on this 😁

**Comment by Tamschi** [comment]
Bridgy Fed maps `_` to `-` since the former isn't allowed either, so that does conflict, unfortunately.
`.` and `-` are also the only characters outside alphanumeric Basic Latin that are valid in ATProto handles.

**Comment by pcottle** [comment]
Well for threads we dont allow `-` in usernames at all, so mapping `pcottle.influencer` to `pcottle-influencer` is fine since no threads user can exist with the `pcottle-influencer` name right?

Just trying to confirm that threads is unblocked, even if the wider issue isnt quite resolved :)

**Comment by Tamschi** [comment]
The problem occurs if you have both `pcottle.influencer` and `pcottle_influencer`.
If you already deny that as confusable then it's all good, though.

(We'd also need some way to make the mapping configurable per instance, maybe.
I think we'd want to keep the `.`s for Bridgy Fed's own translations at least, and give others the option to.)

**Comment by snarfed** [comment]
We're only talking about Bluesky handles here, specifically custom domain handles that Threads might set for its users, right? Not AP actor ids or `preferredUsername` or anything else? If so then this seems fine, Threads can do whatever they want...? Custom Bluesky domain handles are opaque to BF, it doesn't translate to or from them or do anything else with them.

**Comment by pcottle** [comment]
Yep I'm just trying to make sure we can map all usernames over to valid subdomains, without ambiguity or collisions. Ideally without using `.` since thatll be easier to implement.

But this is all unrelated to the topic here (instance opt-in) so Ill take this convo elsewhere :P

**Comment by Tamschi** [comment]
> We're only talking about Bluesky handles here, specifically custom domain handles that Threads might set for its users, right? Not AP actor ids or `preferredUsername` or anything else? If so then this seems fine, Threads can do whatever they want...? Custom Bluesky domain handles are opaque to BF, it doesn't translate to or from them or do anything else with them.

Doesn't Bridgy Fed fail to bridge if it can't assign an initial username by itself or if there's a collision with it?

**Comment by snarfed** [comment]
> Doesn't Bridgy Fed fail to bridge if it can't assign an initial username by itself or if there's a collision with it?

Yes, but if @pcottle is only talking about how Threads would generate Bluesky custom domain handles, that should be unrelated.

**Comment by praichor** [comment]
Unfortunately, opt-in ultimately makes the bridge unusable in practice. As a Mastodon user, I cannot persuade everyone I want to follow on Bluesky to turn on the bridge. In particular, it is impossible to get the attention of celebitry users.

The service is only really useful with opt-out.

And how cool is this bridge!

I also cannot understand the arguments against opt-out: On both sides of the bridge, the basic idea of ​​federation is that the data can be accessed by any other server in the federation, not just current but also future ones. In other words, the data is completely public anyway. You agree to this as soon as you register with a federating social network.

Theoretical question: Would you disapprove, would there a reason against it or would there be anything stopping me from simply hosting a modified opt-out branch of bridgy-fed with a new domain myself? Whoever on Mastodon has something against it can then simply block the domain I use.

**Comment by Fauli1221** [comment]
@praichor this might be the ticket for you #1471

**Comment by praichor** [comment]
> @praichor this might be the ticket for you #1471

Oh yes, this was the wrong ticket. I'll repost in the correct place.

**Comment by S0yKaf** [comment]
> When an instance is ready to try this, I can exempt them from the spam filter req'ts.

Hey, I'm pretty interested in this feature. Would it be possible to exempt `masto.canadiancivil.com` from the filter?
Currently looking into my options in terms of auto adding new users to bridgy-fed.

also, did mastodon remove the auto-follow feature? I know you can use the cli to make accounts follow someone but I can't find the option in the dashboard?

Anyways, seeing as its probably not desirable to bridge a user before they have profile pic I might setup some webhook to wait for that before initiating the bridge.

The 2 week old requirement is also going to have to be removed if the mechanism ends up being officially supported.
I'd argue this probably does more harm then good already. It's not ideal to have a new fediverse user have to wait 2 weeks to have the privilege of interacting with bluesky via bridgy.


Alternatively the relay options that was mentioned earlier seems pretty promising as a way to "activate" the instance opt-in. But just a traditional API could do the trick also.

I'll report back as I get further into my implementation.

**Comment by snarfed** [comment]
@S0yKaf love it! (We've been talking more in https://github.com/snarfed/bridgy-fed/issues/1537#issuecomment-2555739934.)

One catch is that Bridgy Fed doesn't yet support split-domain setups like masto.canadiancivil.com, ie where the server's hostname is different than the webfinger domain in users' handles. Sorry! That's #1031.

**Comment by S0yKaf** [comment]
@snarfed Could you lift the spam filter for `masto.canadiancivil.com` ? I'm ready to start testing my auto-follower implementation.

> One catch is that Bridgy Fed doesn't yet support split-domain setups like masto.canadiancivil.com

What isn't supported exactly? because it seems to work just fine on masto.canadiancivil.com . I did notice that it bridged users using the full instance name instead of the webfinger, but besides that everything seemed to be working properly?

Is the whitelist something that wouldn't work because of the split domain?


Once its tested and working properly I'll most likely release this as a sidecar you can run along your mastodon instance. So it should be pretty painless for anyone who wants to "activate" instance opt-in.

**Comment by snarfed** [comment]
@S0yKaf awesome! I haven't implemented a per-instance bypass for the spam filter, but I can eventually. Until then, feel free to test with existing accounts on your instance that are older than two weeks.

Re split domain support, you're right, the main missing piece is just that it currently uses your instance's hostname in bridged handles, user pages, etc instead of the webfinger domain. Details in #1031. (And split domain support should be unrelated to spam filtering, per-instance bypass there, etc.)

**Comment by snarfed** [comment]
Documented this here: https://fed.brid.gy/docs#enable-api

We're ready for instance admins to try it out! Fully self serve, no need for me to do anything to set you up, just check out the docs and try it. ^

**Comment by trbutler** [comment]
Perhaps I'm being dumb, but the documentation mentions having users follow the Bridgy Fed accounts per normal. Did anyone come up with a way to automate this process on Mastodon? Most of my instance users are not super tech savvy, so making this automatic would be very helpful, but I haven't discovered a built-in way to add two week old accounts to Bridgy Fed without writing some sort of custom tool.

**Comment by grrywlsn** [comment]
> I did notice that it bridged users using the full instance name instead of the webfinger, but besides that everything seemed to be working properly?

Sorry if this has been answered elsewhere, but from the docs and Github issues I haven't found what is expected.

I followed the `Can I give my users custom Bluesky domain handles automatically?` steps in the docs and have the `atproto-did` redirection in place on the domain. Using the Bluesky debug page I can see that it works; everything is green for HTTP verification.

When the instance account follows the bridge account, it correctly shows up in Bluesky. If I search the shortened name in Bluesky, or replace the URL to be `https://bsky.app/profile/username.domain.name`, it correctly shows me the profile.

My question is, is there a way to change the username from appearing as the full `@username.domain.name.ap.brid.gy‬` handle on Bluesky? will the bridge automatically work that out and update at some point in the future?

**Comment by snarfed** [comment]
@grrywlsn thanks for the nudge! You're right, there is one more thing to do to set custom domain handles per user that's not yet in the docs. Sorry for the confusion! I'll add that.

Do you admin a fediverse instance? Are you interested in test driving this? We haven't worked with any admins yet, but we'd love to!

**Comment by trbutler** [comment]
That was for @grrywlsn, but if you're looking for admins to test this, I'd love to help. Several of our users already have manually setup custom domain handles, but I'd love to move to making it an automatic part of the system.

**Comment by grrywlsn** [comment]
> Do you admin a fediverse instance? Are you interested in test driving this? We haven't worked with any admins yet, but we'd love to!

Yes, I have a few instances I can test this with and would be more than happy to.

**Comment by snarfed** [comment]
Awesome! @trbutler @grrywlsn do any of your instances have new users regularly signing up? Those will be more useful for us to test with than small instances with just one or a few users total.

**Comment by grrywlsn** [comment]
Sure - one of the instances is [music-social.com](https://music-social.com), where I create profiles for musical artists so that you can follow and find out about new releases etc. 

I can add as many accounts there as we need to test with. The DNS redirection is already in place.

**Comment by trbutler** [comment]
> Awesome! [@trbutler](https://github.com/trbutler) [@grrywlsn](https://github.com/grrywlsn) do any of your instances have new users regularly signing up? Those will be more useful for us to test with than small instances with just one or a few users total.

We don't have a ton of new user traffic, but there are a few new accounts lined up that need to be added, so I can hold off on those and then add them during the test.

**Comment by snarfed** [comment]
Thanks guys! And are your instances set up to have new users automatically follow @bsky.brid.gy@bsky.brid.gy, as described in https://fed.brid.gy/docs#enable-api ? We don't have a way to detect your new users ourselves, so we need that as a trigger.

**Comment by snarfed** [comment]
Also, do you have a plan for communicating this to new users during the signup flow, or before? Like other differences between instances, we expect that if an instance automatically enables the bridge for new users, it should tell them that loudly when or before they sign up.

**Comment by grrywlsn** [comment]
The accounts on one of my instances automatically follow the bridge, and the accounts are created by me so there's no need to communicate it with other users

**Comment by trbutler** [comment]
> Thanks guys! And are your instances set up to have new users automatically follow @[bsky.brid.gy@bsky.brid.gy](mailto:bsky.brid.gy@bsky.brid.gy), as described in https://fed.brid.gy/docs#enable-api ? We don't have a way to detect your new users ourselves, so we need that as a trigger.

I've been trying to figure out how to implement that in Mastodon without doing a fork of the codebase. Have you found a way? For whatever reason, Mastodon's development team removed the built-in functionality and seem to oppose giving instances the ability to choose to offer accounts with auto-followed accounts -- there are a number of bug reports filed on that, including [this one](https://github.com/mastodon/mastodon/issues/18042).

There's the ability to add a local account to all users via the Mastodon CLI, so I was considering some sort of cron job to run that, but that would break a user's ability to opt-out.  And, of course, Bridgy isn't a local user, so that point is moot.

I'm continuing to research this to see if there's an efficient way to do it. Down the road, it seems like the most user friendly option on an instance that opts-in would be a checkbox in a user's Mastodon account that said "Share with Bluesky," akin to Thread's "Enable Fediverse Sharing," but that's obviously more complex than anything we're discussing here and if upstream Mastodon is resistant, it seems like that would create a maintenance challenge.

@grrywlsn Are you running vanilla Mastodon and getting it to auto follow somehow? Maybe I'm missing something... if you have a tip from getting your instance setup, I'd be grateful!

**Comment by snarfed** [comment]
@trbutler all good questions!

The short answer is, we haven't done a deep dive into how to to do this on each individual fediverse server. We probably should at some point, at least for bigger ones like Mastodon. I'm a bit reluctant to maintain our own copies of docs for other software, and try to keep them updated as those projects change, but it may still be worth it.

For Mastodon specifically, I think this used to be called "default follows" https://github.com/mastodon/mastodon/issues/16229 , and at some point they changed it to follow "suggestions" for new users, https://fedi.tips/recommending-accounts-to-follow-for-new-users-on-your-mastodon-server/ . I don't know a ton about admining Mastodon myself though.

You can also do it lower level, eg https://blog.gcn.sh/mindnotes/is-there-a-way-to-force-users-to-follow-a-mastodon-instance-account

**Comment by snarfed** [comment]
We're also open to a different way to do this that's not follow based! We'd still need some way for instances to tell Bridgy Fed each time a new user signs up, though. Afaik fediverse servers don't generally publish feeds of new users, and I'm not particularly interested in an unofficial polling or scraping based way.

**Comment by snarfed** [comment]
OK @grrywlsn! I see that you have the `/.well-known/atproto-did` redirect set up. Bridgy Fed should now automatically give new users on music-social.com custom domain handles when they enable the bridge. Feel free to try it, let me know how it goes!

**Comment by trbutler** [comment]
Thanks, @snarfed! 

> For Mastodon specifically, I think this used to be called "default follows" [mastodon/mastodon#16229](https://github.com/mastodon/mastodon/issues/16229) , and at some point they changed it to follow "suggestions" for new users, https://fedi.tips/recommending-accounts-to-follow-for-new-users-on-your-mastodon-server/ . I don't know a ton about admining Mastodon myself though.

Yes, that's the problem for trying to implement opt-out. The old default follows would do what I think the goal is of this "instance opt-in" approach. The new suggested follows, less so, since it just displays a list of users someone could follow when they sign up, but knowing to follow Bridgy requires technical savvy since a new user may not immediately understand that this user is key to their Bluesky-using friends being able to follow them. This is the issue in https://github.com/mastodon/mastodon/issues/16229 that you referenced. 

> You can also do it lower level, eg https://blog.gcn.sh/mindnotes/is-there-a-way-to-force-users-to-follow-a-mastodon-instance-account

This was the command I referenced above -- the trouble is, it only works for local accounts, not remote ones like the Bridgy Fed account.

Some admins have been asking Mastodon to add Default Follows back after they removed it a few years ago. Perhaps I should open a new ticket there referencing Bridgy's new functionality as a reason to consider reenabling that. I get their concern about users getting opted into things, but your approach seems quite fair with giving users warning that joining the instance includes a bridge to Bluesky.

**Comment by grrywlsn** [comment]
> OK [@grrywlsn](https://github.com/grrywlsn)! I see that you have the `/.well-known/atproto-did` redirect set up. Bridgy Fed should now automatically give new users on music-social.com custom domain handles when they enable the bridge. Feel free to try it, let me know how it goes!

I tried with a few recent accounts (since the 7 day account age kicks in), but it seems that when they show up in Bluesky they are _not_ given the custom short domain.

For example, the [username in the URL is short](https://bsky.app/profile/tylerthecreator.music-social.com) but the username handle is still the `ap.brid.gy` one.

**Comment by grrywlsn** [comment]
> [@grrywlsn](https://github.com/grrywlsn) Are you running vanilla Mastodon and getting it to auto follow somehow? Maybe I'm missing something... if you have a tip from getting your instance setup, I'd be grateful!

My server is just vanilla Mastodon, that's right; but I'm running a script that interacts with the accounts to both follow the bridge and post as the account.

I think even if the "default follows" feature is re-enabled, the bridge account will still complain since the account is less than 7 days old.

**Comment by snarfed** [comment]
@grrywlsn hmm, we'd exempted music-social.com from the 7 day age req't yesterday when I pinged you.

I'm traveling for the next 2w, but I'll try to look sometime soon and let you know if I see anything obvious.

**Comment by snarfed** [comment]
@grrywlsn figured out the bug, thanks for the nudge. Fix should be deployed in ~20m, feel free to try again then.

**Comment by trbutler** [comment]
> My server is just vanilla Mastodon, that's right; but I'm running a script that interacts with the accounts to both follow the bridge and post as the account.

Would you be willing to share the script? If you'd rather not post it publicly, my e-mail is tbutler@ofb.biz, though I'd imagine it'd be helpful to a lot of folks, since this is an appealing feature in Bridgy Fed. I've been trying to figure out how to accomplish this efficiently with the Mastodon API, covering these issues:

- Acquire the auth code to work as each user in the API.
- Add the account as the user
- Somehow either only add it to users created since the last run or at least avoid readding the follow if a user unfollows.

**Comment by grrywlsn** [comment]
> [@grrywlsn](https://github.com/grrywlsn) figured out the bug, thanks for the nudge. Fix should be deployed in ~20m, feel free to try again then.

All good, thanks - I can see that one of the new accounts that was hitting the 7 day block before got allowed to follow the bridge, and the bridge account has the short domain name handle. 

In terms of existing accounts, is there a way to update their handles? if they unfollow and refollow the bridge, will that work?

**Comment by snarfed** [comment]
> In terms of existing accounts, is there a way to update their handles? if they unfollow and refollow the bridge, will that work?

For existing accounts, you can log into the web UI as them, or DM the Bridgy Fed bot account: https://fed.brid.gy/docs#bluesky-enhanced

**Comment by trbutler** [comment]
OK, I think I have a functional Ruby script that adds users -- my plan is to run it on cron once every 10 minutes; it'll attempt to follow the Bridgy account on any user's account that hasn't already been flagged as processed (preventing a refollow if the user opts out). 

@snarfed  I believe I may have rate limited our instance, though: I initially tried to add all the existing users that hadn't already followed Bridgy Fed and some were added, others are getting an "unexpected error." Do we need FaithTree.social to be exempted by your spam filter to make a bulk add?

Here's the script in case it'd help anyone else. It's perhaps a bit rough -- I'm a Perl guy, so rough on Ruby -- but I believe it works correctly: https://github.com/trbutler/mastodon-autofollow

**Comment by trbutler** [comment]
Second problem: the users that are following Bridgy Fed don't seem to be working with the custom domain. For example: https://trbutler.faithtree.social/.well-known/atproto-did now redirects to https://fed.brid.gy/.well-known/atproto-did?protocol=ap&id=https://FaithTree.social/users/trbutler ... but that redirected page gives 404 Not Found error.

**Comment by snarfed** [comment]
@trbutler hmm! Bridgy Fed itself doesn't (yet) have any rate limits. I haven't set up your domain for custom handles, but I will, and I'll let you know when that's ready.

**Comment by trbutler** [comment]
@snarfed Thanks for enabling that. As to the other issue, I created a new test account so I could watch the process the script initiates and see what it does on the user side, too. (My personal account has been using Bridgy for quite awhile now, so I couldn't test it from there.) I think each of the users who has failed to follow Bridgy are stuck with follow requests. For example:

<img width="525" alt="Image" src="https://github.com/user-attachments/assets/3d913f5a-5a8a-41ec-b375-a89325ac98c7" />

Now, my test account has only existed for a few minutes (not sure if you already suspended the anti-spam measure on my instance), but some of the other accounts that are stuck have been active for more than a year (e.g. @cindib@faithtree.social). Is there any way from Bridgy's end for you to see if the requests are coming in ok? The funny thing is about half of the accounts on the instance that weren't already using Bridgy followed it and about half failed to make the successful follow.

**Comment by trbutler** [comment]
Oh! I found the issue: Bridgy is rejecting the user for lacking a profile picture. Is there a way to disable that check for users? If the users are being added when they join the instance, it seems like that check could get in the way.

<img width="505" alt="Image" src="https://github.com/user-attachments/assets/e7360f1b-56cf-451b-ad4c-9ceb1baf23b6" />

**Comment by snarfed** [comment]
@trbutler yes, turning on Bluesky custom handles also exempts you from the spam checks.

I've added your domain, it should be deployed in ~20m. Your actor ids are on mastodon.faithtree.social though, not faithtree.social - eg see the id field in https://browser.pub/https://faithtree.social/users/trbutler - so you'll need to switch your `/.well-known/atproto-did` redirect to use mastodon.faithtree.social in its id parameter.

I'd also recommend that the domain in the id parameter be all lower case, no capital F or T like in the example you gave. Bridgy Fed should ideally always normalize domains to lower case, but I'm not 100% sure we do that correctly in all cases right now.

**Comment by snarfed** [comment]
Damn, sorry, I take it back, that deploy failed . I'll keep you posted.

**Comment by snarfed** [comment]
OK! Ready now.

**Comment by trbutler** [comment]
Thanks, @snarfed. I made the change to the redirect and the Bluesky seems to be recognizing the custom name for my user (@trbutler.faithtree.social) after momentarily saying I had an invalid handle (since the did was previously authenticated via DNS instead per the non-automatic Bridgy Fed instructions). Some of the other users still get 404, but I assume that's because they generally haven't successfully followed Bridgy yet.

The big question is why they won't follow Bridgy. Take the user @biblicle@faithtree.social that I created to test with earlier today. I've tried to both programmatically to follow Bridgy and following via Mastodon's UI. I verified that an entry goes into Mastodon's follow_requests db briefly until a worker attempts to follow Bridgy and then it vanishes. Is there any way for me to debug from Bridgy's side why the follow request isn't working? 

If it helps, here's the users on the instance and their current status with Bridgy when I run my "auto add" script I linked to above:

* Already following: trbutler
* Already following: littlehillschurch
* Already following: ofb
* Already following: sombragris
* Failed to follow the account for Bigdad1211
* Skipped unapproved user johnk73
* Already following: faithtreecf
* Already following: Joyce
* Failed to follow the account for kaylacat134
* Already following: TammieB
* Already following: Melanie
* Failed to follow the account for CindiB
* Failed to follow the account for JackieLHC
* Failed to follow the account for jeffreylandis
* Failed to follow the account for JKrenning
* Already following: Kathy
* Already following: eryanhaffner
* Failed to follow the account for biblicle


Another curious issue: the accounts that _could_ follow Bridgy and were added as Bridgy followers yesterday (e.g. Kathy, Melanie, TammieB) do not work with the did redirect. For example, https://melanie.faithtree.social/.well-known/atproto-did . The only exception, I believe, is sombragris, though that is imperfect: [Bridgy's web interface](https://fed.brid.gy/ap/@sombragris@faithtree.social) and Bluesky both still use the non-custom domain for the user even though Bluesky will bring the profile up with "@sombragris.faithtree.social."

Thanks for your help!

**Comment by snarfed** [comment]
@trbutler looks like `@biblicle` originally didn't work because it was less than a week old. Your accounts are exempt from that check now, so if you unfollow and re-follow, that should hopefully work.

For accounts that were already bridged before, we aren't yet set up to change their handle from the original one to the custom one. We can take that as a feature request!

**Comment by snarfed** [comment]
> For accounts that were already bridged before, we aren't yet set up to change their handle from the original one to the custom one. 

 ..er, to clarify, we're aren't set up to do it _automatically._ Your can do it manually for them like usual, https://fed.brid.gy/docs#bluesky-enhanced . Just DM or log in and use the UI; your users already have the HTTPS handle resolution set up via the redirect.

**Comment by trbutler** [comment]
@snarfed Something seems to have gone sideways in the process. If I try to get @biblicle to follow Bridgy, it always fails -- even if using the Mastodon web interface, it'll appear to follow Bridgy but then I suppose it fails in the attempt and after a short period goes back to offering to follow it rather than marking it as followed. It _was_ able to follow it yesterday. I'm not seeing any errors in my Mastodon logs, which made me wonder if Bridgy was for some reason denying the follow request?

As to other users:

* sombragris and tammieb: I programmatically initiated a DM from the user to alert Bridgy to the custom user name, but that didn't seem to affect the user's state on Bridgy.

* melanie, joyce and kathy: appears to have successfully followed Bridgy, but the Bridgy web interface says "ap user @[username]@faithtree.social not found"

* trbutler, eryanhaffner, ofb, littlehillschurch, faithtreecf -- these already had Bridgy and custom domain configured prior to trying to do instance opt-in.

* Bigdad1211, kaylacat134, CindiB, JackieLHC, jeffreylandis, JKrenning, Biblicle - fail when sending a follow request to @bsky.brid.gy@bsky.brid.gy.

**Comment by trbutler** [comment]
I turned on debug level logging on Mastodon and only see any errors related to Bridgy that seem to be entirely unrelated to the present issue (perhaps someone trying to follow an account on the server from Bluesky?). So, I created another user (@ bridgeTest@faithtree.social)  and tried both programmatically and manually following Bridgy from that account. The anti-spam error about lacking a profile picture was triggered, so I'm thinking perhaps the issues arising are still somehow tied to the instance not being entirely whitelisted, perhaps?

<img width="624" alt="Image" src="https://github.com/user-attachments/assets/af901757-84d6-4147-a7ad-f1546f28bb89" />

A curious point, though is that while following Bridgy triggered that error, the Bridgy account shows as not being presently followed, so it's as if the error triggered Bridgy to deny the follow request or something?

I tried following your @ snarfed@snarfed.org account from the "BridgeTest" user and it shows as followed, but doesn't show up on [my test account's Follows page](https://mastodon.faithtree.social/@bridgeTest/following). Conversely, I was able to follow local Mastodon users, a remote Mastodon user, a Threads.net user and a Flipboard user, so as far as I can tell FaithTree.social is able to follow different sorts of ActivityPub accounts OK still, there just seems to be a problem with initiating follows with Bridgy since attempting to add instance opt-in.

By the way, I'm not sure if all of this is germane to this ticket since it is about utilizing instance opt-in or if it would be better to open a new ticket about problems with instance opt-in. Feel free to advise. I don't want to spam this ticket!

**Comment by snarfed** [comment]
@trbutler sorry this is still so difficult! Thanks for all your patience here. I'm still not sure what's up between your instance and Bridgy Fed, but I've nudged things on our end, I think your users following the bot should more reliably enable the bridge for them now.

And I'm ok with talking here on this issue, at least for now. Thanks for checking though!

**Comment by trbutler** [comment]
@snarfed thank you so much! I'm very grateful for your help -- and excited about seeing the result! That change you made resulted in a huge difference! The 9 accounts that had been unable to follow Bridgy now are able to -- so that's very encouraging! 

I did notice a lingering issue that I have a feeling might explain why some of those users still do not have a custom handle. For example, TammieB registered with that camelcase username. Mastodon does not appear to be case sensitive in resolving that, but this may be one of those places where Bridgy's system is, as you noted, case sensitive. I switched the NGINX redirect to use `$http_host` instead of `$host` to preserve case in forwarding the result to Bridgy, but users formatted that way still seem to be failing to acquire their custom handles and I suspect having this be case sensitive is undesirable anyway.

(I noticed bsky-debug.app normalizes handles to all lowercase, so when I try to verify one of the users with camelcase names, it makes their username all lowercase and then fails since Bridgy is case sensitive.)

    mastodon@juniper:~/live$ curl https://CindiB.faithtree.social/.well-known/atproto-did -i
    HTTP/2 301 
    server: nginx/1.22.1
    date: Thu, 26 Jun 2025 17:56:09 GMT
    content-type: text/html
    content-length: 169
    location: https://fed.brid.gy/.well-known/atproto-did?protocol=ap&id=https://mastodon.faithtree.social/users/CindiB

But, on bsky-debug, the case sensitive formatting [gets normalized to lowercase](https://bsky-debug.app/handle?handle=CindiB.faithtree.social).

Incidentally, here's the NGINX config I set for the wildcard *.faithtree.social, in case anyone else would benefit from it:

    # Redirect subdomains (except www and mastodon) to user profile
    location / {
        set $subdomain "";
        if ($host ~* "^([a-zA-Z0-9_-]+)\.faithtree\.social$") {
            set $subdomain $1;
        }

        if ($subdomain != "") {
            return 301 https://mastodon.faithtree.social/@$subdomain;
        }

        # Fallback to 404 if invalid subdomain
        return 404;
    }

    # Bridgy Fed atproto-did redirect
    location /.well-known/atproto-did {
        set $subdomain "";
        if ($http_host ~ "^([a-zA-Z0-9_-]+)\.faithtree\.social$") {
            set $subdomain $1;
        }
        access_log /var/log/nginx/atproto-did.log combined;
        if ($subdomain = "") {
            return 400 "Invalid subdomain";
        }
        rewrite ^/.well-known/atproto-did$ https://fed.brid.gy/.well-known/atproto-did?protocol=ap&id=https://mastodon.faithtree.social/users/$subdomain permanent;
    }

**Comment by snarfed** [comment]
@trbutler glad the bridging is working now! And thanks for the info on case sensitivity for handles. ATProto handles are indeed case sensitive; from https://atproto.com/specs/handle#handle-identifier-syntax :

> Handles are not case-sensitive, and should be normalized to lowercase (that is, normalize ASCII A-Z to a-z)

We're on family vacation last week and this coming week, but I can look into this more eventually. I'll file a new issue for the handles.

Thanks also for the nginx config, I may borrow it and add it to our docs!

**Comment by snarfed** [comment]
Looked at this a bit more. Interestingly Mastodon doesn't lower case usernames in its AP ids, it keeps them sensitives. So for your example users above, [kaylacat134](https://browser.pub/https://mastodon.faithtree.social/users/kaylacat134) is `https://mastodon.faithtree.social/users/kaylacat134`, [CindiB](https://browser.pub/https://mastodon.faithtree.social/users/cindib) is `https://mastodon.faithtree.social/users/CindiB`. But ATProto handle resolution will always make its HTTP requests in lower case, eg `https://cindib.faithtree.social/.well-known/atproto-did`.

So, we need a mapping from lower case username to case sensitive username somewhere. I don't currently have that mapping in BF, but I could start building it. Alternatively, the instances could do it themselves, eg redirect `https://cindib.faithtree.social/.well-known/atproto-did` (lower case) to `https://fed.brid.gy/.well-known/atproto-did?protocol=ap&id=https://mastodon.faithtree.social/users/CindiB` (case sensitive).
